Hakawai/Mentions/HKWMentionsStartDetectionStateMachine.m

```diff
@@ -202,22 +209,46 @@ - (void)characterTyped:(unichar)c asInsertedCharacter:(BOOL)inserted previousCha
     }
 }
 
-- (void)deleteTypedCharacter:(unichar)deletedChar withCharacterNowPrecedingCursor:(unichar)precedingChar {
+- (void)deleteTypedCharacter:(unichar)deletedChar
+withCharacterNowPrecedingCursor:(unichar)precedingChar
+                    location:(NSUInteger)location
+                textViewText:(NSString *)textViewText {
     // Determine the type of the character
     enum CharacterType {
         CharacterTypeSeparator = 0,
-        CharacterTypeNormal
+        CharacterTypeNormal,
+        CharacterTypeControlCharacter
     };
     enum CharacterType deletedCharacterType = ([HKWMentionsStartDetectionStateMachine.separatorSet characterIsMember:deletedChar] || deletedChar == 0)
     ? CharacterTypeSeparator
     : CharacterTypeNormal;
-    enum CharacterType currentCharacterType = ([HKWMentionsStartDetectionStateMachine.separatorSet characterIsMember:precedingChar] || precedingChar == 0)
-    ? CharacterTypeSeparator
-    : CharacterTypeNormal;
+    __strong __auto_type delegate = self.delegate;
 
+    // Determine the type of the character
+    enum CharacterType currentCharacterType = CharacterTypeNormal;
+    if ([HKWMentionsStartDetectionStateMachine.whitespaceSet characterIsMember:precedingChar] || precedingChar == 0) {
+        currentCharacterType = CharacterTypeSeparator;
+    } else {
+        // Check first for control character because punctuation character can be control character.
+        // e.g "@" and "#" are punctuation char and might be control character.
+        NSCharacterSet *const controlCharacterSet = [delegate controlCharacterSet];
+        if (controlCharacterSet && [controlCharacterSet characterIsMember:precedingChar]) {
+            currentCharacterType = CharacterTypeControlCharacter;
+        } else if ([HKWMentionsStartDetectionStateMachine.punctuationSet characterIsMember:precedingChar]) {
+            currentCharacterType = CharacterTypeSeparator;
+        }
+    }
     switch (self.state) {
         case HKWMentionsStartDetectionStateQuiescentReady: {
-            if (currentCharacterType == CharacterTypeNormal) {
+            // If user deletes white-space between control character and word, then query mention with word next to whitepace.
+            if (currentCharacterType == CharacterTypeControlCharacter && deletedCharacterType == CharacterTypeSeparator) {
+                self.state = HKWMentionsStartDetectionStateCreatingMention;
+                NSString *const keyword = [HKWMentionsStartDetectionStateMachine wordAfterLocation:location + 1 text:textViewText];
```

Comment:
if you're doing +1s and -1s you should be absolutely sure you're within the index bounds